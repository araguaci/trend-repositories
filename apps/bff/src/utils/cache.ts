const LRU = require('lru-cache');

export let cache = null;

export const initCache = () => {
  const options = {
    max: 500,
    ttl: 86400 * 1000, // 1 day
  };

  cache = new LRU(options);
};
