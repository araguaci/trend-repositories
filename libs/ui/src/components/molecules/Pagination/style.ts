import styled from 'styled-components';

export const StyledPagination = styled.div`
  display: flex;
  justify-content: center;
  padding: 10px;

  > a {
    margin-right: 10px;
  }
`;
