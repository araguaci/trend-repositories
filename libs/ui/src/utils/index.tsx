/* istanbul ignore file */
import { ThemeProvider } from 'styled-components';
import theme from '../theme';
import { ReactNode } from 'react';
import { render, act } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';
import { MockedResponse } from '@apollo/client/testing/core';
import { InMemoryCache } from '@apollo/client/cache';

export const renderWithProvider = (node: ReactNode) => {
  return render(<ThemeProvider theme={theme}>{node}</ThemeProvider>);
};

export const renderWithMockProvider = (
  node: ReactNode,
  cache: InMemoryCache,
  mocks: ReadonlyArray<MockedResponse>
) => {
  return render(
    <ThemeProvider theme={theme}>
      <MockedProvider cache={cache} mocks={mocks}>
        {node}
      </MockedProvider>
    </ThemeProvider>
  );
};

export const renderWithThemeProvider = (node: ReactNode) => {
  return render(<ThemeProvider theme={theme}>{node}</ThemeProvider>);
};

export const waitForResponse = async () =>
  await act(async () => await new Promise((resolve) => setTimeout(resolve, 0)));
export * from '@testing-library/react';
