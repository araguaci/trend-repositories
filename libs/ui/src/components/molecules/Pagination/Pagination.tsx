import { StyledPagination } from './style';
import { Button } from '../../atoms';
import { useRouter } from 'next/router';
import { FC, useEffect, useState } from 'react';

const Pagination: FC<{ total: number }> = ({ total }) => {
  const router = useRouter();

  const [page, setPage] = useState(1);

  useEffect(() => {
    if (router?.query['page']) {
      setPage(+router.query['page']);
    }
  }, [router?.query]);

  const totalPage = Math.ceil(total / 20);
  const previous = page === 1 ? page : page - 1;
  const next = page + 1 < totalPage ? page + 1 : page;

  return (
    <StyledPagination>
      <Button
        data-cy="previous-page"
        disabled={previous === 1 && page === 1}
        href={{
          pathname: '/',
          query: { ...router.query, page: previous },
        }}
      >
        {`< Prev`}
      </Button>

      <Button
        data-cy="next-page"
        disabled={next === totalPage && page === totalPage}
        href={{
          pathname: '/',
          query: { ...router.query, page: next },
        }}
      >
        {`Next >`}
      </Button>
    </StyledPagination>
  );
};

export default Pagination;
