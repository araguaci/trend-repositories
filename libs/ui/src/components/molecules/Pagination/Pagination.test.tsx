import mockRouter from 'next-router-mock';
import singletonRouter from 'next/router';
import { renderWithThemeProvider, screen, fireEvent } from '@ui/utils';
import Pagination from './Pagination';

describe('RepositoryInformation Component Test', () => {
  beforeEach(() => {
    mockRouter.setCurrentUrl('/?page=1');
  });

  it('should be rendered correctly', () => {
    const { asFragment } = renderWithThemeProvider(<Pagination total={100} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('should be previous button disabled', () => {
    // Arrange && Act
    renderWithThemeProvider(<Pagination total={100} />);

    const previous = screen.getByRole('link', { name: /< Prev/ });

    // Assertion
    expect(previous).toHaveStyle('cursor: not-allowed');
  });

  it('should be previous button enabled when page greater then 1', () => {
    // Arrange && Act
    renderWithThemeProvider(<Pagination total={100} />);

    const previous = screen.getByRole('link', { name: /< Prev/ });
    const next = screen.getByRole('link', { name: /Next >/ });

    // Act
    fireEvent.click(next);

    // Assertion
    expect(previous).not.toHaveStyle('cursor: not-allowed');
    expect(singletonRouter).toMatchObject({
      asPath: '/?page=2',
      pathname: '/',
      query: { page: '2' },
    });
  });

  it('should be previous button worked', () => {
    // Arrange && Act
    renderWithThemeProvider(<Pagination total={100} />);

    const previous = screen.getByRole('link', { name: /< Prev/ });
    const next = screen.getByRole('link', { name: /Next >/ });

    fireEvent.click(next);
    expect(previous).not.toHaveStyle('cursor: not-allowed');

    // Act
    fireEvent.click(previous);

    // Assertion
    expect(previous).toHaveStyle('cursor: not-allowed');
    expect(singletonRouter).toMatchObject({
      asPath: '/?page=1',
      pathname: '/',
      query: { page: '1' },
    });
  });

  it('should be next button disabled', () => {
    // Arrange && Act
    mockRouter.setCurrentUrl('/?page=5');

    renderWithThemeProvider(<Pagination total={100} />);

    const next = screen.getByRole('link', { name: /Next >/ });

    // Assertion
    expect(next).toHaveStyle('cursor: not-allowed');
  });
});
