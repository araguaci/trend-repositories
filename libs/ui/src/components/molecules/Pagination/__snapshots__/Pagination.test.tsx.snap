// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepositoryInformation Component Test should be rendered correctly 1`] = `
<DocumentFragment>
  <div
    class="style__StyledPagination-sc-7mqc4q-0 loZPVd"
  >
    <a
      class="style__StyledLink-ikww7b-0 fUma-dV style__StyledButton-gn6n84-0 ljPxBu"
      data-cy="previous-page"
      disabled=""
      href="/?page=1"
    >
      &lt; Prev
    </a>
    <a
      class="style__StyledLink-ikww7b-0 fUma-dV style__StyledButton-gn6n84-0 eAVKfN"
      data-cy="next-page"
      href="/?page=2"
    >
      Next &gt;
    </a>
  </div>
</DocumentFragment>
`;
